service/s3/s3crypto/key_provider.go

```diff
@@ -0,0 +1,54 @@
+package s3crypto
+
+import "crypto/rand"
+
+// KeyProvider is used to help generate keys and ivs. This allows for crypto based
+// algorithms or services to be used.
+type KeyProvider interface {
+	EncrypterKeyProvider
+	DecrypterKeyProvider
+}
+
+// EncrypterKeyProvider handles how to generate keys and ivs for the content encryption.
+// This also satifies the `MaterialDescription` and `CipherDataIface` interfaces.
+type EncrypterKeyProvider interface {
```

Comment:
Yea, I was walking down get also generating values if none was called. However, as I started thinking reusing of the client, it could potentially not generate a key or iv as a bug. Further it made the logic a little wonky. So I decided to just separate the two. If that is what you were talking about